### PR TITLE
Merge ??? from trunk

### PR DESCRIPTION
**Pull request by dcsobral**
Added ??? and NotImplementedError because this seemed to be the
opinion of the majority on the "Adding ??? to Predef?" thread on
scala-internals. No review.

This change is pretty harmless, and if it's good enough for 2.10.0, it ought to be good enough for 2.9.2.

### CODE REVIEW

**Comment by ijuma** [comment]
I don't think we should add API to the 2.9.x series without a very good reason. Otherwise people can't develop against 2.9.2 while cross-compiling to older 2.9.x series.

**Comment by dcsobral** [comment]
I see your point, but adding new features in minor releases (which is what Scala's point releases really are) is common practice. Furthermore, only people cross-compiling exclusively across 2.9.x would benefit from _not_ adding it, because as soon as you go to 2.8.x, you loose even backward compatibility. Finally, this is not a method that introduces an essential feature -- much on the contrary, in fact. It can be easily replaced if necessary, and it's main usefulness is aesthetics, teaching material and power point slides.

File: src/library/scala/NotImplementedError.scala [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+
+
+package scala
+
+/** Throwing this exception can be a temporary replacement for a method
+ *  body that remains to be implemented. For instance, the exception is thrown by
+ *  `Predef.???`.
+ */
+final class NotImplementedError(msg: String) extends Error(msg) {
+  def this() = this("an implementation is missing")
```
Comment (viktorklang) [inline src/library/scala/NotImplementedError.scala:18]: You could get the current stack trace and figure out what method is missing implementation

**Comment by ijuma** [comment]
"I see your point, but adding new features in minor releases (which is what Scala's point releases really are) is common practice."

The closest example I can find is Java update releases and no new API is allowed on those (note that this is different from new features, Java update releases can get new HotSpot features for example).

File: src/library/scala/NotImplementedError.scala (hunk 1 above)
Comment (soc) [inline src/library/scala/NotImplementedError.scala:18]: +1

**Comment by odersky** [comment]
The policy so far is that we do want to allow new features in point releases. The difference to major releases is that we guarantee binary compatibility.

**Comment by ijuma** [comment]
Good to know.

**Comment by paulp** [comment]
All kinds of factors weigh against adding non-essentials to 2.9.1.  Let's aim at getting a 2.10 milestone out instead.

**Comment by dcsobral** [comment]
I'm concentrating mainly on documentation. It's my understanding that 2.9.2 will be out before 2.10 -- if that's not true, then these pull requests are much less meaningful. But if 2.9.2 will be out first, then my reasons are:
- When it comes to documentation, it is impossible to recover any time lost. Whatever damage is inflicted by bad documentation will stay with us.
- ??? is useful as a teaching resource and for slideshow presentations, which is why I opened this pull request as well as the others more directly related to documentation.

**Pull request closed without merging**